rts/lra/lra-coordinator-jar/src/main/java/io/narayana/lra/coordinator/domain/model/Transaction.java

```diff
@@ -228,11 +228,35 @@ private boolean restore_list(InputObjectState os, int ot, RecordList list) {
     public boolean restore_state(InputObjectState os, int ot) {
 
         if (!super.restore_state(os, ot)
-                || !restore_list(os, ot, pendingList)
-                || !restore_list(os, ot, preparedList)) {
+                || !restore_list(os, ot, pendingList)) { // other lists are maintained in BasicAction
             return false;
         }
 
+        // restore_state may have put failed records onto the prepared list so move them back again:
+        for (AbstractRecord rec = preparedList.peekFront(); rec != null; rec = preparedList.peekNext(rec)) {
+            if (rec instanceof LRARecord) {
+                LRARecord p = (LRARecord) rec;
+
+                if (p.isFailed()) {
+                    boolean moveRec = true;
+                    AbstractRecord r;
+
+                    preparedList.remove(p);
+
+                    // put it back on the failedList if it isn't already on it
+                    for (r = failedList.peekFront(); r != null; r = failedList.peekNext(r)) {
```

Comment:
If I'm not mistaken by looking into the code the `failedList` (and neither the `preparedList`) are guarded by some lock or other synchronization guard (while the `RecordList` is not meant to be thread safe). As I'm not fully in-depth into the logic my point is if this code is safe when executed from different threads (if it may happen)?

Reply:
These lists are managed by ArjunaCore and BasicAction - things like save and restore state and commit etc are controlled from BasicAction which have there own synchronisation control.

I can create a JIRA that calls for stress testing of the implementation to try to tease out any dormant issues in this area?

Reply:
I think the stress testing is good but not necessary for the initial insight. If there is the code which can be accessed from different threads and is not synchronized we should be thinking about the circumstances.
As I'm looking into this once again with the understanding the code is invoked from the `BasicAction` I can see that the lists are used in the `Transaction` processing which is the instance of the thread local. That way the processing should be bound to the one thread and the concurrent access of the lists would not be possible.
I"m resolving this doubt here now.